lib/src/dag_walk.rs

```diff
@@ -487,16 +487,31 @@ where
     NI: IntoIterator<Item = Result<T, E>>,
 {
     let start: Vec<T> = start.into_iter().try_collect()?;
-    let mut reachable: HashSet<T> = start.iter().cloned().collect();
-    dfs_ok(start.into_iter().map(Ok), id_fn, |node| {
-        let neighbors: Vec<Result<T, E>> = neighbors_fn(node).into_iter().collect();
-        for neighbor in neighbors.iter().filter_map(|x| x.as_ref().ok()) {
-            reachable.remove(neighbor);
+    let mut heads: HashSet<T> = start.iter().cloned().collect();
+    // Do a BFS until we have only one item left in the frontier. That frontier must
+    // have originated from one of the heads, and since there can't be cycles,
+    // it won't be able to eliminate any other heads.
+    let mut frontier: HashSet<T> = heads.clone();
```

Comment:
nit: `frontier` doesn't have to be `HashSet`, but I'm not sure if earlier deduplication is better or not.

Reply:
Makes sense. Switched to a vector